### PR TITLE
sync the history event from the cluster policy

### PR DESCRIPTION
**Pull request by yanmxa**
Refs: https://issues.redhat.com/browse/ACM-5977

Samples:
```bash
root@myan workspace/hub-of-hubs (br_cluster_policy) » kubectl exec -it $(kubectl get pods -n hoh-postgres -l postgres-operator.crunchydata.com/role=master -o jsonpath='{.items..metadata.name}') -c database -n hoh-postgres -- psql -U postgres -d hoh
psql (14.5)
Type "help" for help.

hoh=# \x auto;
Expanded display is used automatically.
hoh=# select * from event.local_policies;
-[ RECORD 1 ]-+----------------------------------------------------------------------------------------------------------
event_name    | local-placement.policy-limitrange.176cd1b1c6ce0e9a
policy_id     | ba6e4710-878e-4ab0-9131-a1a5b8972ba9
cluster_id    | 69369013-3e0e-4a9c-b38c-7efbe7770b61
leaf_hub_name | kind-hub1
message       | NonCompliant; violation - limitranges not found: [container-mem-limit-range] in namespace default missing
reason        |
count         | 1
source        |
created_at    | 2023-06-28 12:15:25.712544
compliance    | non_compliant
-[ RECORD 2 ]-+----------------------------------------------------------------------------------------------------------
event_name    | local-placement.policy-limitrange.176cd1b02db45911
policy_id     | ba6e4710-878e-4ab0-9131-a1a5b8972ba9
cluster_id    | 69369013-3e0e-4a9c-b38c-7efbe7770b61
leaf_hub_name | kind-hub1
message       | Policy local-placement.policy-limitrange status was updated in cluster namespace kind-hub1-cluster1
reason        | PolicyStatusSync
count         | 1
source        | {"component": "policy-status-sync"}
created_at    | 2023-06-28 12:15:14
compliance    | unknown
-[ RECORD 3 ]-+----------------------------------------------------------------------------------------------------------
event_name    | local-placement.policy-limitrange.176cd1b02db45911
policy_id     | ba6e4710-878e-4ab0-9131-a1a5b8972ba9
cluster_id    | 69369013-3e0e-4a9c-b38c-7efbe7770b61
leaf_hub_name | kind-hub1
message       | Policy local-placement.policy-limitrange status was updated in cluster namespace kind-hub1-cluster1
reason        | PolicyStatusSync
count         | 2
source        | {"component": "policy-status-sync"}
created_at    | 2023-06-28 12:15:21
compliance    | non_compliant

```

### CODE REVIEW

File: agent/pkg/status/bundle/grc/cluster_policy_status_event_bundle.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+	policy, ok := object.(*policiesv1.Policy)
+	if !ok {
+		return // do not handle objects other than policy
+	}
+	bundlePolicyStatusEvents, ok := bundle.PolicyStatusEvents[string(policy.GetUID())]
+	if !ok {
+		bundlePolicyStatusEvents = make([]*statusbundle.PolicyStatusEvent, 0)
+	}
+
+	if policy.Status.Details == nil {
```
Comment (clyang82) [inline agent/pkg/status/bundle/grc/cluster_policy_status_event_bundle.go:61]: move it before creating `bundlePolicyStatusEvents`. if `Details` is nil, should not create `bundlePolicyStatusEvents` to save memory.

File: pkg/bundle/status/base_policy_history_event_bundle.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// ClusterPolicyHistoryEventBundle the base struct for cluster policy history event bundle.
+type BaseClusterPolicyStatusEventBundle struct {
+	LeafHubName        string                            `json:"leafHubName"`
+	PolicyStatusEvents map[string]([]*PolicyStatusEvent) `json:"policyStatusEvents"`
+	BundleVersion      *BundleVersion                    `json:"bundleVersion"`
+}
+
+type PolicyStatusEvent struct {
```
Comment (clyang82) [inline pkg/bundle/status/base_policy_history_event_bundle.go:12]: we may share the same struct defined in database/models/event

File: agent/pkg/status/bundle/grc/cluster_policy_status_event_bundle.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	if err != nil {
+		return
+	}
+
+	modified := false
+	for _, detail := range policy.Status.Details {
+		if detail.History != nil {
+			for _, event := range detail.History {
+				bundlePolicyStatusEvents = bundle.loadEventToBundle(event, detail, eventMap,
+					string(rootPolicy.GetUID()), clusterId, bundlePolicyStatusEvents, &modified)
```
Comment (clyang82) [inline agent/pkg/status/bundle/grc/cluster_policy_status_event_bundle.go:95]: We need to delete the deprecated events from Bundle if it is not in `detail.History`.

**Review by clyang82 — CHANGES_REQUESTED** [review]

File: agent/pkg/status/bundle/grc/cluster_policy_status_event_bundle.go (hunk 3 above)
Comment (yanmxa) [reply agent/pkg/status/bundle/grc/cluster_policy_status_event_bundle.go:98]: Done!

File: pkg/bundle/status/base_policy_history_event_bundle.go (hunk 2 above)
Comment (yanmxa) [reply pkg/bundle/status/base_policy_history_event_bundle.go]: Done!

File: agent/pkg/status/bundle/grc/cluster_policy_status_event_bundle.go (hunk 1 above)
Comment (yanmxa) [reply agent/pkg/status/bundle/grc/cluster_policy_status_event_bundle.go:57]: Done!

**Pull request merged**